Tests/NIOCoreTests/ByteBufferTest.swift

```diff
@@ -1843,19 +1843,108 @@ class ByteBufferTest: XCTestCase {
         60 61 62 63 64 65 66 67 68 69 6a 6b 6c 6d 6e 6f 70 71 72 73 74 75 76 77 78 79 7a 7b 7c 7d 7e 7f \
         80 81 82 83 84 85 86 87 88 89 8a 8b 8c 8d 8e 8f 90 91 92 93 94 95 96 97 98 99 9a 9b 9c 9d 9e 9f \
         a0 a1 a2 a3 a4 a5 a6 a7 a8 a9 aa ab ac ad ae af b0 b1 b2 b3 b4 b5 b6 b7 b8 b9 ba bb bc bd be bf \
         c0 c1 c2 c3 c4 c5 c6 c7 c8 c9 ca cb cc cd ce cf d0 d1 d2 d3 d4 d5 d6 d7 d8 d9 da db dc dd de df \
-        e0 e1 e2 e3 e4 e5 e6 e7 e8 e9 ea eb ec ed ee ef f0 f1 f2 f3 f4 f5 f6 f7 f8 f9 fa fb fc fd fe ff ]
+        e0 e1 e2 e3 e4 e5 e6 e7 e8 e9 ea eb ec ed ee ef f0 f1 f2 f3 f4 f5 f6 f7 f8 f9 fa fb fc fd fe ff
         """
         XCTAssertEqual(expected, actual)
     }
 
+    func testHexDumpPlain() {
+        let buf = ByteBuffer(string: "Hello")
+        XCTAssertEqual("48 65 6c 6c 6f", buf.hexDump(format: .plain))
+    }
+
+    func testHexDumpPlainEmptyBuffer() {
+        let buf = ByteBuffer(string: "")
+        XCTAssertEqual("", buf.hexDump(format: .plain))
+    }
+
+    func testHexDumpPlainWithReaderIndexOffset() {
+        var buf = ByteBuffer(string: "Hello")
+        let firstTwo = buf.readBytes(length: 2)!
+        XCTAssertEqual([72, 101], firstTwo)
+        XCTAssertEqual("6c 6c 6f", buf.hexDump(format: .plain))
+    }
+
+    func testHexDumpPlainWithMaxBytes() {
+        self.buf.clear()
+        for f in UInt8.min...UInt8.max {
+            self.buf.writeInteger(f)
+        }
+        let actual = self.buf.hexDump(format: .plain(maxBytes: 10))
+        let expected = "00 01 02 03 04 ... fb fc fd fe ff"
+        XCTAssertEqual(expected, actual)
+    }
+
+    func testHexDumpDetailed() {
+        let buf = ByteBuffer(string: "Goodbye, world! It was nice knowing you.\n")
+        let expected = """
+        00000000  47 6f 6f 64 62 79 65 2c  20 77 6f 72 6c 64 21 20  |Goodbye, world! |
+        00000010  49 74 20 77 61 73 20 6e  69 63 65 20 6b 6e 6f 77  |It was nice know|
+        00000020  69 6e 67 20 79 6f 75 2e  0a                       |ing you..|
+        00000029
+        """
+        let actual = buf.hexDump(format: .detailed)
+        XCTAssertEqual(expected, actual)
+    }
+
+
+    func testHexDumpDetailedWithMaxBytes() {
+        let buf = ByteBuffer(string: "Goodbye, world! It was nice knowing you.\n")
+        let expected = """
+        00000000  47 6f 6f 64 62 79 65 2c                           |Goodbye,        |
+        ........  .. .. .. .. .. .. .. ..  .. .. .. .. .. .. .. ..  ..................
+        00000020                           6e 67 20 79 6f 75 2e 0a  |        ng you..|
+        00000029
+        """
+        let actual = buf.hexDump(format: .detailed(maxBytes: 16))
+        XCTAssertEqual(expected, actual)
+    }
+
+    func testHexDumpDetailedWithMultilineFrontAndBack() {
+        let buf = ByteBuffer(string: """
+        Goodbye, world! It was nice knowing you.
+        I will miss this pull request with all of it's 94+ comments.
```

Comment:
We can keep it going if you like? 😅 (Also, thank you again, this is a really awesome addition!)

Reply:
such good changes